pkg/virtual/apiexport/controllers/apireconciler/apireconciler_controller.go

```diff
@@ -0,0 +1,287 @@
+
+	apiExportLister  apislisters.APIExportLister
+	apiExportIndexer cache.Indexer
+
+	queue workqueue.RateLimitingInterface
+
+	createAPIDefinition CreateAPIDefinitionFunc
+
+	mutex   sync.RWMutex // protects the map, not the values!
+	apiSets map[dynamiccontext.APIDomainKey]apidefinition.APIDefinitionSet
+}
+
+func (c *APIReconciler) enqueueAPIResourceSchema(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	clusterName, name := clusters.SplitClusterAwareKey(key)
+	exports, err := c.apiExportIndexer.ByIndex(byWorkspace, clusterName.String())
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	for _, obj := range exports {
+		export := obj.(*apisv1alpha1.APIExport)
+		klog.V(2).Infof("Queueing APIExport %s|%s for APIResourceSchema %s", clusterName, export.Name, name)
+		c.enqueueAPIExport(obj)
+	}
+}
+
+func (c *APIReconciler) enqueueAPIExport(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (c *APIReconciler) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *APIReconciler) Start(ctx context.Context) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s controller", ControllerName)
+	defer klog.Infof("Shutting down %s controller", ControllerName)
+
+	go wait.Until(func() { c.startWorker(ctx) }, time.Second, ctx.Done())
+
+	// stop all watches if the controller is stopped
+	defer func() {
+		c.mutex.Lock()
+		defer c.mutex.Unlock()
+		for _, sets := range c.apiSets {
+			for _, v := range sets {
+				v.TearDown()
+			}
+		}
+	}()
+
+	<-ctx.Done()
+}
+
+func (c *APIReconciler) ShutDown() {
+	c.queue.ShutDown()
+}
+
+func (c *APIReconciler) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%s: failed to sync %q, err: %w", ControllerName, key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *APIReconciler) process(ctx context.Context, key string) error {
+	clusterName, apiExportName := clusters.SplitClusterAwareKey(key)
+
+	apiExport, err := c.apiExportLister.Get(key)
+	if err != nil {
+		klog.Errorf("failed to get APIExport %s|%s from lister: %v", clusterName, apiExportName, err)
+		return nil // nothing we can do here
+	}
+	apiDomainKey := dynamiccontext.APIDomainKey(clusterName.String() + "/" + apiExportName)
+
+	if apiExport.Status.IdentityHash == "" {
+		// this APIExport is not ready for consumption by controllers. remove all APIs.
+		c.mutex.Lock()
+		defer c.mutex.Unlock()
```

Comment:
IMO this is harder to read and reason about than

```
lock
delete
unlock
```

Reply:
I usually do defer with locks whenever possible, and your style only when needed (without adding another closure).